Add manual section on modular configurations (#1841)
tweag/nickel: doc/manual/modular-configurations.md

---
slug: modular-configurations
---

# Modular configurations

> This section explains how to leverage the Nickel languages features, and in

Reviewer: ```suggestion
This section explains how to leverage Nickel's features, and in
```

One of the starting points of Nickel has been the limits of static data formats
such as JSON, TOML or YAML when tackling large configurations. In particular,
those formats have no way of reusing values of your configuration. Have a single
value used in several places (say, an IP)? Have this virtual machine description
block used several times, almost identically but for a few parameters? In JSON,
the best you can do is to copy and paste, praying for the different copies to not
drift away with time.

In Nickel, you can just use variables and functions:

```nickel
let master_ip = "192.168.1.23" in
let machine = fun {ip, open_ports} => {
  # machine definition depending on ip and open_ports
} in

{
  ip = master_ip,
  cmd = "ssh -p 8070 user@%{master_ip}",
  client1 = machine {ip = "0.0.0.0", open_ports = [22]},
  client2 = machine {ip = "1.1.1.1", open_ports = [80]},
}
```

In the zoology of programming abstractions, functions are the simplest and most
ubiquitous concept of all. For the machine example, a ([pure][pure-function])
function is exactly what we look for: an expression with unknown parameters
which can be selected at usage time. Virtually every programming language has

> functions in some form, even if they're not called that way (for example, in

Reviewer: ```suggestion
functions in some form, even if they're not called that (for example, in
```

Author: I don't have a very strong opinion, but this part is verbatim from the blog which went under the review of the editing team. Is it a stylistic issue, or a correctness? If the former, I think I'll stick to the published version. If the latter, let me know, because then I need to update the blog post as well.

Reviewer: It feels to me like "called that way" is about how something gets called, e.g. by phone. But I'm not a native speaker, so take that with a grain of salt. I don't think anyone will care, though :)

Terraform's configuration language, _modules_ are more or less
functions[^terraform-modules]).

> You might even want not to hardcode some parameters directly in the

Reviewer: ```suggestion
You might even want to not hardcode some parameters directly in the
```

Author: I think both are valid, although the emphasize is a bit different, right?

Reviewer: I'm not sure; "not want to" or "want to not" would feel right to me and have this difference in emphasis. But again, not a native speaker...

argument.

It also happens that one doesn't control the call-site or the function's code,
because it's in a library. That is, overriding is **opt-in for the author of the
function**, but the one who needs it is the **consumer**. It's hard for the
author to know in advance everything that could be useful to expose for
overriding.

Overriding is typically an issue in Nix: when pulling package descriptions from
the central repository ([Nixpkgs][nixpkgs]), the package is "already applied".
For example, we build Nickel with Nix, and we need to make sure the version of
`wasm-bindgen` we pull from Nixpkgs is the same as the one set in `Cargo.toml`.
A naive way would be to do something like (in Nix, `//` overrides a value in a
record):

```nix
wasm_bindgen_cli =
  pkgs.wasm-bindgen-cli // {version = wasm_bindgen_fixed_version;}
```

Sadly, this doesn't work. This code overrides the final `version` field
correctly, but not the other values which depend on version (such as `src`).
What we want to override is the _original_ `version` used when calling the
package-as-a-function, not the one propagated in the end result. But once an
arbitrary function has been applied, there's no coming back.

> The [actual Nix code][wasm-bindgen-override] to build Nickel calls to a

Reviewer: ```suggestion
The [actual Nix code][wasm-bindgen-override] to build Nickel calls a
```

```console
$ echo '(import "machine.ncl") & {ip = "1.1.1.1"}' > config.ncl
$ nickel export config.ncl
{
  "ip": "1.1.1.1"
  "cmd": "ssh -p 8070 user@1.1.1.1",
}
```

Deep down, recursive records could be seen as just a fancy representation for
functions (in the same way as [objects are poor man's
closures][objects-closures]). But we'll see how this representation makes many
configuration-related operations much easier.

### Records are transparent

As opposed to functions, partial records don't need to hang on their missing
field. We can extract documentation from them, get completion in the LSP, use
`nickel query`, and so on. The structure of the final configuration is already
apparent: there are just some holes to be filled.

### Records remember

As opposed to functions, recursive records remember the relation between their
different fields even if they have all been given a value at this point.
Overriding a field will automatically recompute the new value of its reverse
dependencies. Remember the `wasm-bindgen` example, where we had to use Nixpkgs'
`override`. In Nickel, you don't need any of that. The following is a similar

> example, albeit simplified, where the `cmd` field depends on the the `ip` field

Reviewer: ```suggestion
example, albeit simplified, where the `cmd` field depends on the `ip` field
```

(but, in return, we can then remove the `not_exported`).

All of those values are readily and automatically overridable (using the `force`
metadata for fields that already have a non-default value), and changes are
propagated to reverse dependencies accordingly. The proposed layout is rather
arbitrary: Nickel has no notion of inputs, local and outputs, but rather of
fields with and without definition as well as priorities.

The metadata associated to `inputs` define a clear typed module interface which
can be leveraged by the Nickel tooling.

Nickel might get a first-class notion of modules one day, if only to enforce a
standard structure across projects. But, as far as functionality is concerned,
the current merge system can already get you pretty far in an intuitive way.

## Conclusion

In this section, we explored the merge system of Nickel and how to leverage it
to write natural configuration parts that are easy to query, to combine and to
tweak. We supported our claim that partial configuration is a better suited
model than functions (or other abstractions, for that matter) for configuration
management, and showed that the combination of merging and field metadata is an

> expressive combo.

Reviewer: I feel like we should remove this conclusion. It doesn't really fit the style of the other manual sections.

Author: Fair enough. I wondered about it too, because we have "take-away" subsections in some sections (typing I believe), but I agree it doesn't really fit.

> 
> ## Conclusion
> 
> In this section, we explored the merge system of Nickel and how to leverage it
> to write natural configuration parts that are easy to query, to combine and to
> tweak. We supported our claim that partial configuration is a better suited
> model than functions (or other abstractions, for that matter) for configuration
> management, and showed that the combination of merging and field metadata is an
> expressive combo.

Author: ```suggestion
```
